### PR TITLE
Vc2 updates

### PR DESCRIPTION
**Pull request by brycelelbach**
Patches to get octotiger working with Vc2, which has AVX512 support.

### CODE REVIEW

File: src/simd.hpp [hunk 1]
```diff
@@ -402,7 +402,7 @@ class v4sd {
 #include <hpx/parallel/traits/vector_pack_type.hpp>
 #include <hpx/runtime/serialization/datapar.hpp>
 
-using simd_vector = typename hpx::parallel::traits::vector_pack_type<double, 8>::type;
+using simd_vector = Vc::datapar<double, Vc::datapar_abi::native<double>>;
```
Comment (hkaiser) [inline src/simd.hpp]: Will this always give (i.e. on any platform) a vector size of `8`?

File: src/taylor.hpp [hunk 2]
```diff
@@ -469,15 +469,20 @@ inline void taylor<5, simd_vector>::set_basis(const std::array<simd_vector, NDIM
...
     }
-// #else
-//     where(r2 > 0.0) | r2inv = ONE / r2;
-// #endif
+#else
+    #ifdef Vc_IS_VERSION_2  
+        where(r2 > 0.0, r2inv) = ONE / r2;
+        using Vc::sqrt;
+    #else
+        where(r2 > 0.0) | r2inv = ONE / r2;
```
Comment (hkaiser) [inline src/taylor.hpp]: We found out the hard way, that `where` works on assignment only. The intention was to perform only the _calculation_ for the elements which are non-negative, but as it turns out the code will perform the division for all elements, masking the assignment only.

**Comment by brycelelbach** [comment]
No, you are right. This patch is not right.

The problem is that fixed size doesn't reduce to intrinsics currently and
multipurpose if necessary (e.g. fixed size of 8 on sse2 only hardware isn't
hardwired to do 4x sse2 intrinsics), so the compiler may not optimize them.

I think the best short term solution is to check if we are on avx512 and if
so, go to native, otherwise use fixed.

Also... It may be worth double checking that the optimized intrinsic
version is better. It was in my preliminary testing.

On Mar 20, 2017 4:58 AM, "Hartmut Kaiser" <notifications@github.com> wrote:

*@hkaiser* commented on this pull request.
------------------------------

In src/simd.hpp
<https://github.com/STEllAR-GROUP/octotiger/pull/48#discussion_r106883743>:

> @@ -402,7 +402,7 @@ class v4sd {
 #include <hpx/parallel/traits/vector_pack_type.hpp>
 #include <hpx/runtime/serialization/datapar.hpp>

-using simd_vector = typename
hpx::parallel::traits::vector_pack_type<double, 8>::type;
+using simd_vector = Vc::datapar<double, Vc::datapar_abi::native<double>>;

Will this always give (i.e. on any platform) a vector size of 8?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/STEllAR-GROUP/octotiger/pull/48#pullrequestreview-27825022>,
or mute the thread

**Comment by brycelelbach** [comment]
Are we worried about the cost of duplicating the computation?

**Comment by hkaiser** [comment]
> Are we worried about the cost of duplicating the computation?

No, this is to avoid division by zero errors.

**Comment by brycelelbach** [comment]
Oh. Ouch! I'll send a revised pr today

**Comment by brycelelbach** [comment]
Uh-oh. This is troubling. The code that the Vc version 1 build fails with is #ifdef'd with Vc_IS_VERSION_2... so perhaps we can't actually use this macro to distinguish between version 1 and version 2.

Investigating.

**Comment by brycelelbach** [comment]
I think this is good to go now.

**Pull request merged**